File: src/jianzhi28.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        //要把空节点给考虑进去
        //nice，一次过，十五分钟
        //时间击败：86.79%，空间击败：49.68%
        if(root == nullptr || (root->left == nullptr && root->right == nullptr)){
            return true;
        }
        else{
            bool result = isSymmetric(root->left, root->right); //函数重载
            return result;
        }
    }

    bool isSymmetric(const TreeNode * _node1, const TreeNode * _node2){
        //分情况讨论
        if(_node1 == nullptr && _node2 == nullptr){
            //cout<<"*\n";
            return true;
        }
        else if(_node1 == nullptr || _node2 == nullptr){
            //只有一方为空
            //cout<<"$\n";           
            return false;
        }
        else if(_node1->val != _node2->val){
            //cout<<"&\n";          
            cout<<_node2->val<<" "<<_node2->val<<endl;
            //两个节点都存在，但是数值不等
            return false;
        }
        else{
            
            //cout<<_node2->val<<" "<<_node2->val<<endl;
            //两个节点都存在，并且数值相等，继续向下
            //左对右，右对左
            return isSymmetric(_node1->left, _node2->right) && isSymmetric(_node1->right, _node2->left);
        }
    }
};

File: src/jianzhi28.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        #递归，每次比较两个节点
        #干翻蚂蚁金服
        #五分钟，一次提交，时间击败-80.00%，空间击败-50.35%
        #边界判断
        if root == None:
            return True
        def judgeSymmetric(node1, node2):
            if node1 == None and node2 == None:
                return True
            elif node1 == None:
                return False
            elif node2 == None:
                return False
            elif node1.val != node2.val:
                return False
            else:
                return judgeSymmetric(node1.left, node2.right) and judgeSymmetric(node1.right, node2.left)

        return judgeSymmetric(root.left, root.right)
